lib/utils/parse/parse.go

```diff
@@ -149,20 +160,132 @@ func RoleVariable(variable string) (*Expression, error) {
 	}, nil
 }
 
+// Matcher matches strings against some internal criteria (e.g. a regexp)
+type Matcher interface {
+	Match(in string) bool
+}
+
+// NewMatcher parses a matcher expression. Currently supported expressions:
+// - string literal: `foo`
+// - wildcard expression: `*` or `foo*bar`
+// - regexp expression: `^foo$`
+// - regexp function calls:
+//   - positive match: `{{regexp.match("foo.*")}}`
+//   - negative match: `{{regexp.not_match("foo.*")}}`
+//
+// These expressions do not support variable interpolation (e.g.
+// `{{internal.logins}}`), like Expression does.
+func NewMatcher(value string) (Matcher, error) {
+	match := reVariable.FindStringSubmatch(value)
+	if len(match) == 0 {
+		if strings.Contains(value, "{{") || strings.Contains(value, "}}") {
+			return nil, trace.BadParameter(
+				"%q is using template brackets '{{' or '}}', however expression does not parse, make sure the format is {{expression}}",
+				value)
+		}
+		return newRegexpMatcher(value, true)
+	}
+
+	prefix, variable, suffix := match[1], match[2], match[3]
+
+	// parse and get the ast of the expression
+	expr, err := parser.ParseExpr(variable)
+	if err != nil {
+		return nil, trace.BadParameter("failed to parse %q: %v", variable, err)
+	}
+
+	// walk the ast tree and gather the variable parts
+	result, err := walk(expr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// For now, only support a single match expression. In the future, we could
+	// consider handling variables and transforms by propagating user traits to
+	// the matching logic. For example
+	// `{{regexp.match(external.allowed_env_trait)}}`.
+	if result.transform != nil || len(result.parts) > 0 {
+		return nil, trace.BadParameter("%q is not a valid matcher expression - no variables and transformations are allowed", variable)
+	}
+	return newPrefixSuffixMatcher(prefix, suffix, result.match), nil
+}
+
+// regexpMatcher matches input string against a pre-compiled regexp.
+type regexpMatcher struct {
+	re *regexp.Regexp
+}
+
+func (m regexpMatcher) Match(in string) bool {
+	return m.re.MatchString(in)
+}
+
+func newRegexpMatcher(raw string, escape bool) (*regexpMatcher, error) {
+	if escape {
+		if !strings.HasPrefix(raw, "^") || !strings.HasSuffix(raw, "$") {
+			// replace glob-style wildcards with regexp wildcards
+			// for plain strings, and quote all characters that could
+			// be interpreted in regular expression
+			raw = "^" + utils.GlobToRegexp(raw) + "$"
+		}
+	}
+
+	re, err := regexp.Compile(raw)
+	if err != nil {
+		return nil, trace.BadParameter("failed parsing regexp %q: %v", raw, err)
```

Comment:
May make sense to send a link to our documentation with examples. This will make our message actionable and reduce amount of support.